rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,677 @@
+
+## Additional properties on types
+
+In the proposed [Versioning RFC](./0408-versioning-types.md) for the type system, having `{ "additionalProperties": false }` for all schemas is an assumption made for determining type compatibility, which means that any supertype will not validate against a subtype that adds properties if it receives all properties of a subtype instance. For example, if we supply the `Employee` instance from above to a `Person`, it will receive properties that are considered `additionalProperties` (the `Occupation` property is not present on `Person`).
+
+The assumption that we can select/project parts of a subtype that make up a supertype is essential for keeping strictness in JSON Schemas.
+
+And unfortunately, specifying `{ "unevaluatedProperties": false }` does not behave as expected when composing types together in JSON Schema either, which means we will have to redefine how `{ "additionalProperties": true }` or `{ "unevaluatedProperties": false }` behaves within our type extension system to make supertypes keep strictness while allowing composition.
+
+Concrete examples of how JSON Schema breaks with these validation constraints are shown in the [Reference-level explanation](#additional-properties-on-types-1)
+
+## Defining extended types
+
+Extended types will be defined with conventional JSON Schema syntax, the `allOf` keyword. When creating a new entity type it's possible to extend another entity type by adding an entry to `allOf` value with a versioned URI reference.
+Using a versioned URI makes it so that subtypes aren't automatically updated when the supertype is.
+
+As extended types can extend other extended types, we must also make sure that there are no cycles within the type hierarchy, as it could lead to hard to reason about types and unpredictability.
+
+## Addressing previous considerations
+
+In the [Type System RFC](./0352-graph-type-system.md#interfacing-with-types-1), we made the following statement:
+
+> The update messages of both the current and new systems make use of partial schemas, merging the schema given in the message contents with the existing Entity Type. This may or may not be the desired semantics of updating, and could lead to undesired behavior. In that case, the semantics can be changed to treat updates as a complete replacement. This is to be decided and can be considered out of scope for this RFC as it touches on inheritance/forking concepts.
+
+As the new Block Protocol type system doesn't require a delta-based storage approach for its schemas, it's unclear what the advantages of partial schemas for update requests are. We propose that we should treat type updates as complete replacements, so implementation is much more straightforward for embedding applications. Partial schema updates also add some level of indirection, and that may obfuscate error sources and error reasons. Therefore, updates to types must be complete replacements rather than partial schema updates.
```

Comment:
> I think there may be EAs that want to be able to control access to certain property types on an entity type (not talking about the entity data here), to avoid leaking the structure of protected data, etc.

I get that controlling access might be beneficial here, but how would a situation where a user has access to modify an entity type but not have read access on the fields look like? If a user does not have access to certain property types, meaning that the block on a user's page doesn't either, how would a block know of the property type missing on an entity type in the first place?

Reply:
I more meant that imagine you have some entity type
```json
{
  "id": "http://..../Employee",
  "properties": {
    "financial stuff": ...
    "hr-y stuff": ...
  }
```

_Perhaps_ we enter a world where accounting users can access the "financial stuff" property on Employee, and have update permissions on "financial stuff", but cannot see the HR stuff, and vice versa for HR users. 

But again I guess this is super hypothetical at this point and:
> But perhaps that's something they can work around in the implementation anyway (e.g. it's a "full replacement" of the part being sent)

Meaning maybe the accounting users are only exposed to a "view" of the Employee where they don't even know there's missing properties, so it's still a "full replace" of the projection.

I think we can ignore my comment for now :)

Reply:
Can this comment be resolved?

Reply:
Can this comment be resolved @benwerner01?